### PR TITLE
New Android suppport

### PR DESCRIPTION
**Pull request by larsbergstrom**
r/f? @mbrubeck 

No need to r+ urgently; I want to do a little bit more testing of the release build, but I'm hoping to land this bit (moving to a more sane build process) next week.

The new version of building an APK:
1) Removes the glutin-based APK builder from the link step
2) Adds a build.rs step to the build of the final Servo library that adds the native code required by glutin's android_rs_glue (e.g., `ANativeActivity_onCreate` definition)
3) Replaces the link step with a `fake-ld.sh` script that instead creates a libservo.so
4) Adds a new mach `package` step to build the APK that has some Rust code that builds the library from a set of in-tree build files

This setup fixes a number of problems:
1) We can use gdb, because we use `ndk-build`, which adds the .gdbserver info, plus we keep around all of the build files (also required by the ndk gdb)
2) We can add more Java code & hooks to handle Android intents
3) We no longer have any git submodules or the awkward two-step build with android-rs-glue

Many other setups were tried (and failed). The most obvious ones is building a libservo.so from a `dylib` target from the servo build on Android. This doesn't work because you can't have a different default lib target on one platform than others in Cargo, and you also can't pass it in from the commandline (e.g., --lib does not have a dylib arg). Additionally, if you don't go through the intermediate libservo.rlib step (which removes unused symbols), then you end up with a TON of missing symbols because our -sys crates are super sloppy about that. I spent a few weeks beginning to clean them up, but since it's something we can't easily enforce (and new -sys packages will have this problem, too, since it's only an issue with the Android loader), it made more sense to me to just have the build set up to discard those unused bits of code before they ever get to the linker, much less the loader.

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/8288)

### CODE REVIEW

**Comment by eefriedman** [comment]
Review status: 0 of 24 files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[.gitmodules, line 3 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8288#-K1zZndW0PYThhb3YkzS)** ([raw file](https://github.com/servo/servo/blob/53d8f04ac4894480aa8bc7a6a79fb5e02ae050d1/.gitmodules#L3)):</sup>
Did you mean to delete this file?


---

<sup>**[components/servo/build.rs, line 11 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8288#-K1zZt6d3bGRJKY0SfVq)** ([raw file](https://github.com/servo/servo/blob/db410f2d27ab090a4863389e24393c31fcc1dc7d/components/servo/build.rs#L11)):</sup>
If might be a bit more clear to separate this into two functions: main() and android_main().


---

<sup>**[support/android/build-apk/Cargo.toml, line 15 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8288#-K1zZcNwzbbdTvXz1f_-)** ([raw file](https://github.com/servo/servo/blob/db410f2d27ab090a4863389e24393c31fcc1dc7d/support/android/build-apk/Cargo.toml#L15)):</sup>
Leftover comment?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8288)

**Comment by larsbergstrom** [comment]
Review status: 0 of 24 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

<sup>**[components/servo/build.rs, line 11 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8288#-K1zZt6d3bGRJKY0SfVq)** ([raw file](https://github.com/servo/servo/blob/db410f2d27ab090a4863389e24393c31fcc1dc7d/components/servo/build.rs#L11)):</sup>
Will do.


---

<sup>**[support/android/build-apk/Cargo.toml, line 15 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8288#-K1zZcNwzbbdTvXz1f_-)** ([raw file](https://github.com/servo/servo/blob/db410f2d27ab090a4863389e24393c31fcc1dc7d/support/android/build-apk/Cargo.toml#L15)):</sup>
Thanks! I'll get it.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8288)

**Comment by larsbergstrom** [comment]
@eefriedman re "Did you mean to delete this file?"

Yup, I'm remove the last git submodule in our tree - woo-hoo!

**Comment by eefriedman** [comment]
Err, I meant "Did you mean to delete this file, as opposed to leaving it in the tree as a zero-byte file".

**Comment by larsbergstrom** [comment]
Aha, then yes, I'll delete it - that was the intent :-)

**Comment by eefriedman** [comment]
From tidy: `./python/servo/post_build_commands.py:202: W391 blank line at end of file`

**Comment by larsbergstrom** [comment]
BTW, definitely don't merge this; there's a small linker problem on Android release-only where it's stripping some of the critical public symbols (e.g., `ANativeActivity_onCreate`) that I still need to figure out :-/

File: components/servo/build.rs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+use std::env;
+use std::path::Path;
+use std::process;
+use std::process::{Command, Stdio};
+
+fn main() {
+    // build.rs is not platform-specific, so we have to check the target here.
+    let target = env::var("TARGET").unwrap();
+    if !target.contains("android") {
+        return;
```
Comment (eefriedman) [inline components/servo/build.rs]:
```
if target.contains("android") {
    android_main()
}
```

**Comment by eefriedman** [comment]
(This looks fine in terms of style, but I don't know enough about how building for Android works to review this properly.)

**Comment by larsbergstrom** [comment]
@mbrubeck With that last fixup! (thanks to @tomaka for the hint!), we have working debug and release packages - I think this is ready to be reviewed and landed, when you have a chance.

**Comment by mbrubeck** [comment]
Reviewed 17 of 24 files at r1.
Review status: 13 of 25 files reviewed at latest revision, 5 unresolved discussions.

---

<sup>**[python/servo/post_build_commands.py, line 185 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8288#-K2Dg4K6zdIcGJcgDD4j)** ([raw file](https://github.com/servo/servo/blob/db410f2d27ab090a4863389e24393c31fcc1dc7d/python/servo/post_build_commands.py#L185)):</sup>
It would be nice if this used the same logic as `CommandBase.get_binary_path` to determine the default if neither `dev` nor `release` is set.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8288)

**Comment by larsbergstrom** [comment]
Review status: 13 of 25 files reviewed at latest revision, 5 unresolved discussions.

---

<sup>**[python/servo/post_build_commands.py, line 185 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8288#-K2Dg4K6zdIcGJcgDD4j)** ([raw file](https://github.com/servo/servo/blob/db410f2d27ab090a4863389e24393c31fcc1dc7d/python/servo/post_build_commands.py#L185)):</sup>
Hrm, I might have to pull out the code in `get_binary_path` into something separate that determines if it's dev or release, then, b/c I need to know for how to set the env vars. Does that sound OK?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8288)

**Comment by larsbergstrom** [comment]
Review status: 13 of 25 files reviewed at latest revision, 5 unresolved discussions.

---

<sup>**[components/servo/build.rs, line 14 \[r3\]](https://reviewable.io:443/reviews/servo/servo/8288#-K1zV8UdV41_NbGCsk0j-r3-14)** ([raw file](https://github.com/servo/servo/blob/382661cd6111791bd604815a2ea309964d8980bc/components/servo/build.rs#L14)):</sup>
Fixed


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8288)

**Comment by larsbergstrom** [comment]
Review status: 13 of 25 files reviewed at latest revision, 5 unresolved discussions.

---

<sup>**[.gitmodules, line 3 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8288#-K1zZndW0PYThhb3YkzS)** ([raw file](https://github.com/servo/servo/blob/53d8f04ac4894480aa8bc7a6a79fb5e02ae050d1/.gitmodules#L3)):</sup>
Fixed; deleted now.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8288)

**Comment by mbrubeck** [comment]
Reviewed 6 of 24 files at r1.
Review status: 18 of 25 files reviewed at latest revision, 6 unresolved discussions.

---

<sup>**[python/servo/post_build_commands.py, line 185 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8288#-K2Dg4K6zdIcGJcgDD4j)** ([raw file](https://github.com/servo/servo/blob/db410f2d27ab090a4863389e24393c31fcc1dc7d/python/servo/post_build_commands.py#L185)):</sup>
> Hrm, I might have to pull out the code in get_binary_path into something separate

Yes, that sounds good.


---

<sup>**[support/android/apk/src/rust/glutin/MainActivity.java, line 1 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8288#-K2DjI_I-LtW_Q-HJPnQ)** ([raw file](https://github.com/servo/servo/blob/db410f2d27ab090a4863389e24393c31fcc1dc7d/support/android/apk/src/rust/glutin/MainActivity.java#L1)):</sup>
Any reason for this package name in particular?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8288)

**Comment by larsbergstrom** [comment]
Review status: 18 of 25 files reviewed at latest revision, 6 unresolved discussions.

---

<sup>**[support/android/apk/src/rust/glutin/MainActivity.java, line 1 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8288#-K2DjI_I-LtW_Q-HJPnQ)** ([raw file](https://github.com/servo/servo/blob/db410f2d27ab090a4863389e24393c31fcc1dc7d/support/android/apk/src/rust/glutin/MainActivity.java#L1)):</sup>
Oh! That was just a carry-over. I should make it `com.mozilla.servo` like I did with the default android package name. Good catch :-)


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8288)

**Comment by larsbergstrom** [comment]
@mbrubeck OK, set to go (modulo having to open a bunch of E-Easy issues <grin>)

**Comment by mbrubeck** [comment]
Reviewed 1 of 25 files at r1, 5 of 5 files at r2, 1 of 1 files at r3, 2 of 2 files at r4, 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

<sup>**[support/android/build-apk/src/main.rs, line 99 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8288#-K2JYwZDL9qCW9Hb9GH_)** ([raw file](https://github.com/servo/servo/blob/db410f2d27ab090a4863389e24393c31fcc1dc7d/support/android/build-apk/src/main.rs#L99)):</sup>
File a follow-up to reduce the duplication of this value?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8288)

**Comment by mbrubeck** [comment]
r=mbrubeck, needs squash

**Comment by larsbergstrom** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

<sup>**[support/android/build-apk/src/main.rs, line 99 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8288#-K2JYwZDL9qCW9Hb9GH_)** ([raw file](https://github.com/servo/servo/blob/db410f2d27ab090a4863389e24393c31fcc1dc7d/support/android/build-apk/src/main.rs#L99)):</sup>
Duplication of which value? The line you commented on is the "android-18", but I only see it there once. Or do you mean that the target SDK platform should be passed in to build-apk from mach?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8288)

**Comment by mbrubeck** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

<sup>**[support/android/build-apk/src/main.rs, line 99 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8288#-K2JYwZDL9qCW9Hb9GH_)** ([raw file](https://github.com/servo/servo/blob/db410f2d27ab090a4863389e24393c31fcc1dc7d/support/android/build-apk/src/main.rs#L99)):</sup>
The string "android-18" is hard-coded in three different files. It'd be nice to have it configured in one place.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8288)

**Comment by jdm** [comment]
@bors-servo: retry
* #8311

**Pull request merged**